src/ray/common/memory_monitor.cc

```diff
@@ -0,0 +1,112 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <tuple>
+#include "ray/common/memory_monitor.h"
+#include <fstream>      // std::ifstream
+#include "ray/util/logging.h"
+#include "ray/common/ray_config.h"
+
+
+namespace ray {
+
+MemoryMonitor::MemoryMonitor(float capacity_threshold,
+    uint64_t monitor_interval_ms,
+    MemoryUsageRefreshCallback monitor_callback)
+  : usage_threshold_(capacity_threshold),
+    monitor_callback_(monitor_callback),
+    io_context_(),
+    monitor_thread_([this] {
+      boost::asio::io_service::work io_service_work_(io_context_);
+      io_context_.run();
+    }),
+    runner_(io_context_) {
+  if (monitor_interval_ms > 0) {
+    #ifdef __linux__
+      runner_.RunFnPeriodically([this] { 
+          bool is_usage_above_threshold = IsUsageAboveThreshold();
+          monitor_callback_(is_usage_above_threshold);
+        },
+        monitor_interval_ms,
+        "MemoryMonitor.CheckIsMemoryUsageAboveThreshold");
+      RAY_LOG(INFO) << "MemoryMonitor initialized";
+    #else
+      RAY_LOG(WARNING) << "Not running MemoryMonitor. It is currently supported on Linux.";
+    #endif    
+  } else {
+    RAY_LOG(INFO) << "MemoryMonitor disabled";
+  }
+}
+
+
+bool MemoryMonitor::IsUsageAboveThreshold() {
+  auto [available_memory_bytes, total_memory_bytes]
+          = GetNodeAvailableMemoryBytes();
+  auto used = total_memory_bytes - available_memory_bytes;
+  auto usage_fraction = static_cast<float>(used) / total_memory_bytes;
```

Comment:
would this cause divid by 0 issue if the total_memory_bytes is 0?

Reply:
Good catch - fixed